README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

[![CRAN
Version](http://www.r-pkg.org/badges/version/soundgen)](https://cran.r-project.org/package=soundgen)
[![Downloads](http://cranlogs.r-pkg.org/badges/soundgen)](https://CRAN.R-project.org/package=soundgen)

R package for sound synthesis and acoustic analysis.  
Homepage with help, demos, etc: <http://cogsci.se/soundgen.html>  
Source code on github: <https://github.com/tatters/soundgen>

Performs parametric synthesis of sounds with harmonic and noise
components such as animal vocalizations or human voice. Also includes
tools for spectral analysis, pitch tracking, audio segmentation,
self-similarity matrices, morphing, etc.

# Key functions

  - Sound synthesis from R console: `soundgen()`
  - Shiny app for sound synthesis (opens in a browser): `soundgen_app()`
  - Acoustic analysis of a wav/mp3 file / folder: `analyze()` / `analyzeFolder()`
  - Shiny app for editing intonation contours (opens in a browser):
    `pitch_app()`
  - Shiny app for measuring formants (opens in a browser):
    `formant_app()`
  - Measuring syllables, pauses, and bursts in a wav/mp3 file / folder:
    `segment()` / `segmentFolder`

For more information, please see the vignettes on sound synthesis and
acoustic analysis:

`vignette("sound_generation", package="soundgen")`

`vignette("acoustic_analysis", package="soundgen")`

Or, to open the vignettes in a browser:  
`RShowDoc('sound_generation', package = 'soundgen')`

`RShowDoc('acoustic_analysis', package = 'soundgen')`

# Example of sound synthesis

Use the `soundgen()` function to create a breathy moan:

``` r
s = soundgen(
  sylLen = 230,               # duration of voiced part, ms
  pitch = c(305, 280),        # pitch, Hz (goes down from 305 to 380 Hz)
  ampl = c(0, -20),           # amplitude, dB (gradual fade-out by 20 dB)
  rolloff = -30,              # strong f0, weak harmonics
  temperature = 0.05,         # some stochasticity in generation
  formants = c(260, 960, 1500, 2200,  # F1-F8 formant frequencies, Hz
               2600, 3600, 4200, 4500),
  noise = data.frame(
    time = c(-50, 120, 700),  # time of noise anchors
    value = c(-15, -5, -50)   # noise amplitude, dB
  ),
  rolloffNoise = 0,           # flat noise spectrum before adding formants
  addSilence = 0, samplingRate = 44100, pitchSamplingRate = 44100,
  play = TRUE, plot = TRUE, osc = TRUE, ylim = c(0, 6)
)
```

<img src="man/figures/README-synthsesis-1.png" width="80%" style="display: block; margin: auto;" />

<p>
</p>

# Example of acoustic analysis

Use the `analyze()` function to detect pitch and obtain other spectral
descriptives of the sound we have just synthesized:

``` r
a = analyze(s, 44100, ylim = c(0, 6))
#> Scale not specified. Assuming that max amplitude is 1
```

<img src="man/figures/README-analysis-1.png" width="80%" style="display: block; margin: auto;" />

``` r
a[1:5, c('pitch', 'peakFreq', 'harmHeight', 'HNR', 'ampl', 'loudness')]
#>      pitch peakFreq harmHeight       HNR      ampl  loudness
#> 1       NA       NA         NA        NA 0.0305833        NA
#> 2       NA 305.2747         NA  1.083205 0.1442435  9.636858
#> 3 300.0882 305.2747   1197.802 12.934929 0.3642711 17.808124
#> 4 299.4505 305.2747   1197.802 14.042033 0.5131389 22.289716
#> 5 296.0563 305.2747   1110.114 13.450248 0.4959925 22.193244
colnames(a)
#>  [1] "duration"           "duration_noSilence" "time"               "ampl"              
#>  [5] "amplVoiced"         "dom"                "entropy"            "f1_freq"           
#>  [9] "f1_width"           "f2_freq"            "f2_width"           "f3_freq"           
#> [13] "f3_width"           "harmEnergy"         "harmHeight"         "HNR"               
#> [17] "loudness"           "peakFreq"           "pitch"              "quartile25"        
#> [21] "quartile50"         "quartile75"         "specCentroid"       "specSlope"         
#> [25] "voiced"
```

# Installation

To install the current release from CRAN: `install.packages("soundgen")`

NB: Make sure all dependencies have been installed correctly\! For
problems with seewave, see <http://rug.mnhn.fr/seewave/>

On Macs, you may need to do the following:

  - First install brew according to the instructions here:
    <https://brew.sh/>
  - Then run the following from the terminal  
    `brew install libsndfile`  
    `brew install fftw`
  - Finally, install soundgen in R:  
    `install.packages("soundgen")`


NEWS
          CHANGES IN SOUNDGEN VERSION 2.1.0.9000 [--date--]
NEW FUNCTIONS:
  * shiftPitch(): shifts pitch and/or formants + performns time stretch
  * resample(): changes the sampling rate without introducing artefacts like aliasing
  * bandpass(): low-pass, high-pass, bandpass, and bandstop filters with FFT

MINOR
  * shiftFormants(): arguments "scaleFactor" renamed to "multFormants" to be consistent with the new function shiftPitch()


          CHANGES IN SOUNDGEN VERSION 2.1.0 [19-June-2021]
NEW FUNCTIONS:
  * shiftFormants(): raises or lowers formants in a sound without changing its pitch
  * pitchDescriptives(): extracts common descriptives of time series such as pitch contours: measures of average / range / variability / slope / inflections etc
  * findInflections(): finds inflections in time series such as pitch contours

MINOR
  * analyze(), modulationSpectrum(): new outputs "amFreq" and "amDep" give the frequency and depth of amplitude modulation
  * analyze(), pitch_app(): adaptive second-pass priors (new argument "priorAdapt")
  * analyze(), pitch_app(): prior distribution for pitch candidates is lognormal instead of gamma
  * switch from Seeing Speech to Hillenbrand 1995 for vowel centroids in formant_app()
  * getSmoothContour(): downsampling is preceeded by low-pass filtering to avoid aliasing
  * formant_app(): adjustable playback speed
  * pitch_app(): support for opening previously saved pitch contours (added as manual anchors)
  * reportTime() and all loops that use it: max two time levels shown (eg days + hours)

BUGS FIXED:
  * analyze(), pitch_app(): pitch contour plotted as both lines and points to ensure that isolated voiced frames surrounded by NAs are shown
  * formant_app(): vowel plot, VTL updated in output table, synthetic audio playback with browser, NAs not filled in user annotations
  * formant_app(), pitch_app(): bug-prone "overlap" setting removed, use "step" instead
  * pitch_app(): brush is cleared on double click instead of "mouseout" (otherwise it's hard to operate buttons like voice/unvoice)
  * ssm(): silence has novelty = 0
  * analyze(): correctly plots extraContour with negative values (eg HNR, amDep)


          CHANGES IN SOUNDGEN VERSION 2.0.0 [15-March-2021]
This version represents an attempt to standardize and better integrate all functions for acoustic analysis in the package. Functions for sound synthesis, notably soundgen(), are not affected by these structural changes and remain fully back-compatible.

GLOBAL (seriously affects back-compatibility):
  * Functions analyzeFolder(), spectrogramFolder(), segmentFolder(), modulationSpectrumFolder(), getRMSFolder(), getLoudnessFolder() are deprecated.
  * Instead, analyze(), spectrogram(), osc(), segment(), modulationSpectrum(), getRMS(), getLoudness(), ssm(), playme(), etc. accept as input either a single sound (numeric vector, Wave object, or filename) or multiple sounds (a list of vectors or Wave objects, multiple filenames like "c(file1, file2)"", or path to folder with audio files).
  * Functions for acoustic analysis return lists with both detailed results and summaries; for example, analyze() returns a list with two components: "$detailed" (frame-by-frame acoustic measures, one dataframe per file) and "$summary" (summaries based on "summaryFun", one row per file).
  * Harmonized terminology throughout the package for basic arguments like "from", "to", saving plots ("savePlots") and audio ("saveAudio"), reporting estimated time left ("reportEvery"), etc.

MAJOR (affects back-compatibility)
  * segment(): a new segmentation algorithm based on spectral signal-noise separation, several deprecated and several new arguments, support for long files (>1 h of audio), an option to export detected syllables as separate files, etc.
  * transplantFormants(): a single argument "freqWindow", support for Wave input, etc.
  * analyze(): SPL_measured and Pref moved to a new argument "loudness = list(SPL_measured, Pref, spreadSpectrum)"
  * analyze(): interpolWin, interpolTol, interpolCert moved to a new argument "interpol = list(win, tol, cert)"
  * compareSounds(): multivariate DTW, different input formats, new arguments, etc.

NEW FUNCTIONS:
  * reverb(): adds reverberation and/or echo to a sound
  * getEnv(): returns the amplitude envelope of a sound
  * audSpectrogram(): produces a spectrogram by applying a bank of bandpass filters with Mel-spaced central frequencies (experimental)

MINOR:
  * spectrogram(), analyze(): a new argument "extraContour" for plotting something over the spectrogram
  * analyze(): spectral slope (specSlope) is returned in dB/kHz
  * analyze() calls ssm() and returns a novelty contour (a new argument "novelty")
  * analyze(): updated cepstral pitch tracking, new argument in pitchCep = list(cepPenalty) to regulate how much high-frequency cepstral candidates are mistrusted
  * analyze() returns strength of subharmonics per frame (output "subRatio" and "subDep")
  * pitch_app(), formant_app(): user zoom settings preserved when moving between audio files in queue
  * formant_app(): added a plot of relative formant frequencies to compare with IPA vowels
  * soundgen(): pitch anchors can contain NAs, eg "pitch = c(150, 200, NA, 110)" - the 3rd quarter will be unvoiced
  * addFormants(): new argument "zFun" for arbitrary transformations of the spectrogram between STFT and iSTFT (eg for de-noising, spectral warping, etc.)
  * flatEnv() AKA compressor(): a new argument "compression" regulates the amount of compression, updated plotting
  * ssm(): a new argument "sparse" for quickly extracting just the novelty contour from long sounds, changes in default settings, input "audiogram" renamed to "melspec", etc.

BUGS FIXED:
  * formant_app(): fixed occasional crashing when scrolling with left/right buttons, other small-scale debugging

R/morph.R
#' Morph sounds
#'
#' Takes two formulas for synthesizing two target sounds with
#' \code{\link{soundgen}} and produces a number of intermediate forms (morphs),
#' attempting to go from one target sound to the other in a specified number of
#' equal steps. Normally you will want to set \code{temperature} very low; the
#' \code{tempEffects} argument is not supported.
#'
#' @seealso \code{\link{soundgen}}
#'
#' @param formula1,formula2 lists of parameters for calling
#'   \code{\link{soundgen}} that produce the two target sounds between which
#'   morphing will occur. Character strings containing the full call to soundgen
#'   are also accepted (see examples)
#' @param nMorphs the number of morphs to produce, including target sounds
#' @param playMorphs if TRUE, the morphs will be played
#' @param savePath if it is the path to an existing directory, morphs will be
#'   saved there as individual .wav files (defaults to NA)
#' @param samplingRate sampling rate of output, Hz. NB: overrides the values in
#'   \code{formula1} and \code{formula2}
#' @return A list of two sublists (\code{$formulas} and \code{$sounds}), each
#'   of length \code{nMorphs}. For ex., the formula for the second hybrid is
#'   \code{m$formulas[[2]]}, and the waveform is \code{m$sounds[[2]]}
#' @export
#' @examples
#' # write two formulas or copy-paste them from soundgen_app() or presets:
#' playback = c(TRUE, FALSE)[2]
#' # [a] to barking
#' m = morph(formula1 = list(repeatBout = 2),
#'           # equivalently: formula1 = 'soundgen(repeatBout = 2)',
#'           formula2 = presets$Misc$Dog_bark,
#'           nMorphs = 5, playMorphs = playback)
#'  # use $formulas to access formulas for each morph, $sounds for waveforms
#'  # m$formulas[[4]]
#'  # playme(m$sounds[[3]])
#'
#' \dontrun{
#' # morph intonation and vowel quality
#' m = morph(
#'   'soundgen(pitch = c(300, 250, 400),
#'             formants = c(350, 2900, 3600, 4700))',
#'   'soundgen(pitch = c(300, 700, 500, 300),
#'             formants = c(800, 1250, 3100, 4500))',
#'   nMorphs = 5, playMorphs = playback
#' )
#'
#' # from a grunt of disgust to a moan of pleasure
#' m = morph(
#'   formula1 = 'soundgen(sylLen = 180, pitch = c(160, 160, 120), rolloff = -12,
#'     nonlinBalance = 70, subDep = 15, jitterDep = 2,
#'     formants = c(550, 1200, 2100, 4300, 4700, 6500, 7300),
#'     noise = data.frame(time = c(0, 180, 270), value = c(-25, -25, -40)),
#'     rolloffNoise = 0)',
#'   formula2 = 'soundgen(sylLen = 320, pitch = c(340, 330, 300),
#'     rolloff = c(-18, -16, -30), ampl = c(0, -10), formants = c(950, 1700, 3700),
#'     noise = data.frame(time = c(0, 300, 440), value = c(-35, -25, -65)),
#'     mouth = c(.4, .5), rolloffNoise = -5, attackLen = 30)',
#'   nMorphs = 8, playMorphs = playback
#' )
#'
#' # from scream_010 to moan_515b
#' # (see online demos at http://cogsci.se/soundgen/humans/humans.html)
#' m = morph(
#'   formula1 = "soundgen(
#'     sylLen = 490,
#'     pitch = list(time = c(0, 80, 250, 370, 490),
#'     value = c(1000, 2900, 3200, 2900, 1000)),
#'     rolloff = c(-5, 0, -25), rolloffKHz = 0,
#'     temperature = 0.001,
#'     jitterDep = c(.5, 1, 0), shimmerDep = c(5, 15, 0),
#'     formants = c(1100, 2300, 3100, 4000, 5300, 6200),
#'     mouth = c(.3, .5, .6, .5, .3))",
#'   formula2 = "soundgen(sylLen = 520,
#'     pitch = c(300, 310, 300),
#'     ampl = c(0, -30),
#'     temperature = 0.001, rolloff = c(-18, -25),
#'     jitterDep = .05, shimmerDep = 2,
#'     formants = list(f1 = c(700, 900),
#'       f2 = c(1600, 1400),
#'       f3 = c(3600, 3500), f4 = c(4300, 4200)),
#'     mouth = c(.5, .3),
#'     noise = data.frame(time = c(0, 400, 660),
#'     value = c(-20, -10, -60)),
#'     rolloffNoise = c(-5, -15))",
#'  nMorphs = 5, playMorphs = playback
#' )
#' }
morph = function(formula1,
                 formula2,
                 nMorphs,
                 playMorphs = TRUE,
                 savePath = NA,
                 samplingRate = 16000) {
  # convert formulas to lists
  if (is.character(formula1)) {
    formula1 = paste0('list', substr(formula1, 9, nchar(formula1)))
    formula1 = eval(parse(text = formula1))
  }
  if (!is.list(formula1)) {
    stop('Formula1 must be either a list of pars like "list(sylLen = 500)"
         or a character string like "soundgen(sylLen = 500"')
  }

  if (is.character(formula2)) {
    formula2 = paste0('list', substr(formula2, 9, nchar(formula2)))
    formula2 = eval(parse(text = formula2))
  }
  if (!is.list(formula2)) {
    stop('Formula2 must be either a list of pars like "list(sylLen = 500)"
         or a character string like "soundgen(sylLen = 500"')
  }

  # expand all abbreviations to full format
  if (!is.null(formula1$formants)) {  # otherwise drops NULL completely
    formula1$formants = reformatFormants(formula1$formants)
  }
  if (!is.null(formula2$formants)) {
    formula2$formants = reformatFormants(formula2$formants)
  }

  if (!is.null(formula1$formantsNoise)) {
    formula1$formantsNoise = reformatFormants(formula1$formantsNoise)
  }
  if (!is.null(formula2$formantsNoise)) {
    formula2$formantsNoise = reformatFormants(formula2$formantsNoise)
  }

  for (anchor in c('pitch', 'pitchGlobal', 'glottis',
                   'ampl', 'amplGlobal', 'mouth')) {
    if(!is.null(formula1[[anchor]])) {
      formula1[[anchor]] = reformatAnchors(formula1[[anchor]])
    }
    if(!is.null(formula2[[anchor]])) {
      formula2[[anchor]] = reformatAnchors(formula2[[anchor]])
    }
  }
  if (is.numeric(formula1$noise)) {
    if (length(formula1$noise) > 0) {
      formula1$noise = data.frame(
        time = seq(0,
                   ifelse(is.numeric(formula1$sylLen),
                          formula1$sylLen,
                          defaults$sylLen),
                   length.out = max(2, length(formula1$noise))),
        value = formula1$noise)
    }
  }
  if (is.numeric(formula2$noise)) {
    if (length(formula2$noise) > 0) {
      formula2$noise = data.frame(
        time = seq(0,
                   ifelse(is.numeric(formula2$sylLen),
                          formula2$sylLen,
                          defaults$sylLen),
                   length.out = max(2, length(formula2$noise))),
        value = formula2$noise)
    }
  }

  # set pitchSamplingRate to the higher value
  formula1$samplingRate = formula2$samplingRate = samplingRate
  if (is.numeric(formula1$pitchSamplingRate) |
      is.numeric(formula2$pitchSamplingRate)) {
    mps = max(formula1$pitchSamplingRate, formula2$pitchSamplingRate)
    formula1$pitchSamplingRate = formula2$pitchSamplingRate = mps
  }

  # ascertain that the formulas are not identical
  if (identical(formula1[order(names(formula1))], formula2[order(names(formula2))])) {
    stop('Nothing to morph!')
  }

  # which pars are different from the defaults of soundgen()?
  defaults = as.character(call('print', args(soundgen)))[2]
  defaults = substr(defaults, 11, (nchar(defaults) - 12))
  defaults = paste0('list(', defaults, ')')
  defaults = eval(parse(text = defaults))
  # NB: defaults = as.list(args(soundgen)) produces list elements of type "call" :((
  # exclude booleans
  defaults = defaults[!names(defaults) %in% c('plot', 'play', 'savePath', 'invalidArgAction')]

  notDefaultIdx_formula1 = notDefaultIdx_formula2 = NA
  if (length(formula1) > 0) {
    notDefaultIdx_formula1 = which(apply(matrix(1:length(formula1)), 1, function(x) {
      identical(formula1[[x]], defaults[[names(formula1)[x]]]) == FALSE
    }))
  }
  if (length(formula2) > 0) {
    notDefaultIdx_formula2 = which(apply(matrix(1:length(formula2)), 1, function(x) {
      identical(formula2[[x]], defaults[[names(formula2)[x]]]) == FALSE
    }))
  }

  # these pars have to be morphed:
  notDefaultNames = unique(c(names(formula1)[notDefaultIdx_formula1],
                             names(formula2)[notDefaultIdx_formula2]))
  notDefaultNames1 = names(formula1) [names(formula1) %in% notDefaultNames]
  notDefaultNames2 = names(formula2) [names(formula2) %in% notDefaultNames]

  # set up two formulas that contain the same number of pars, fill in with
  # specified values or, for values that are not specified, with defaults
  f1 = f2 = defaults[notDefaultNames]
  f1[notDefaultNames1] = formula1[match(notDefaultNames1, names(formula1))]
  f2[notDefaultNames2] = formula2[match(notDefaultNames2, names(formula2))]
  f2 = f2[match(names(f1), names(f2))]

  suppressWarnings({
    # fill in formant stuff if it is missing for one sound but defined for the other
    # b/c defaults contain formants in c() form, repeat:
    if (!is.null(f1$formants)) {  # otherwise drops NULL completely
      f1$formants = reformatFormants(f1$formants)
    }
    if (!is.null(f2$formants)) {
      f2$formants = reformatFormants(f2$formants)
    }

    if ((!is.list(f1$formants) & is.list(f2$formants)) |
        (is.list(f1$formants) & !is.list(f2$formants))) {
      stop(paste('Specify formants either for both formulas or for neither'))
    }

    if ((!is.list(f1$formantsNoise) & is.list(f2$formantsNoise)) |
        (is.list(f1$formantsNoise) & !is.list(f2$formantsNoise))) {
      stop(paste('Specify formantsNoise either for both formulas or for neither'))
    }

    if ((any(is.na(unlist(f1$formants))) & !any(is.na(unlist(f2$formants)))) |
        (!any(is.na(unlist(f1$formants))) & any(is.na(unlist(f2$formants))))) {
      stop(paste('All formant amplitudes and bandwidths should be either left default',
                 'or fully specified in both formulas, not a mix'))
    }

    if ((any(is.na(unlist(f1$formantsNoise))) & !any(is.na(unlist(f2$formantsNoise)))) |
        (!any(is.na(unlist(f1$formantsNoise))) & any(is.na(unlist(f2$formantsNoise))))) {
      stop(paste('All formantsNoise amplitudes and bandwidths should be either left default',
                 'or fully specified in both formulas, not a mix'))
    }
  })

  # NULL noise means these anchors should be set to the default -dynamicRange
  if ('noise' %in% notDefaultNames) {
    if (!is.numeric(f1$noise)) {
      f1$noise = reformatAnchors(-defaults$dynamicRange)
    }
    if (!is.numeric(f2$noise)) {
      f2$noise = reformatAnchors(-defaults$dynamicRange)
    }
  }

  # log-transform pitch and formant frequencies before morphing
  if ('pitch' %in% names(f1)) {
    if (is.numeric(f1$pitch$value)) f1$pitch$value = log(f1$pitch$value)
    if (is.numeric(f2$pitch$value)) f2$pitch$value = log(f2$pitch$value)
  }
  if ('formants' %in% names(f1)) {
    for (l in 1:length(f1$formants)) {
      if (is.numeric(f1$formants[[l]]$freq)) {
        f1$formants[[l]]$freq = log(f1$formants[[l]]$freq)
      }
    }
    for (l in 1:length(f2$formants)) {
      if (is.numeric(f2$formants[[l]]$freq)) {
        f2$formants[[l]]$freq = log(f2$formants[[l]]$freq)
      }
    }
  }

  if ('formantsNoise' %in% names(f1)) {
    for (l in 1:length(f1$formantsNoise)) {
      if (is.numeric(f1$formantsNoise[[l]]$freq)) {
        f1$formantsNoise[[l]]$freq = log(f1$formantsNoise[[l]]$freq)
      }
    }
    for (l in 1:length(f2$formantsNoise)) {
      if (is.numeric(f2$formantsNoise[[l]]$freq)) {
        f2$formantsNoise[[l]]$freq = log(f2$formantsNoise[[l]]$freq)
      }
    }
  }

  # expand attackLen
  if (length(formula1$attackLen) == 2 & length(formula2$attackLen) < 2) {
    formula2$attackLen = rep(formula2$attackLen, 2)
  }
  if (length(formula2$attackLen) == 2 & length(formula1$attackLen) < 2) {
    formula1$attackLen = rep(formula1$attackLen, 2)
  }

  # expand vectorized pars, if any, to anchors
  for (i in 1:length(f1)) {
    if (is.numeric(f1[[i]]) & is.numeric(f2[[i]])) {
      if (length(f1[[i]]) > 1 | length(f2[[i]]) > 1) {
        f1[[i]] = reformatAnchors(f1[[i]])
        f2[[i]] = reformatAnchors(f2[[i]])
      }
    }
  }

  # f1 and f2 are now fully prepared: two target formulas,
  # both of the same length, including the same pars, in the same order

  # MORPH THE FORMULAS
  m = f1
  formulas = rep(list(f1), nMorphs)
  sounds = list()
  for (p in 1:length(f1)) {
    # morph each element of the formula list according to its type
    if (class(f1[[p]])[1] == 'numeric') {
      m[[p]] = seq(f1[[p]], f2[[p]], length.out = nMorphs)
    } else if (names(f1[p]) %in% c('formants', 'formantsNoise')) {
      m[[p]] = morphList(f1[[p]], f2[[p]], nMorphs = nMorphs)
    } else {
      m[[p]] = morphDF(f1[[p]], f2[[p]], nMorphs = nMorphs)
    }
    # a convoluted way of saving the output of morphFormants() in appropriate
    # slots in the output list
    for (h in 1:nMorphs) {
      formulas[[h]] [[p]] = m[[p]] [[h]]
    }
  }
  # END OF MORPHING THE FORMULAS

  # exponentiate pitch and formant frequencies after morphing
  if ('pitch' %in% names(f1)) {
    for (h in 1:nMorphs) {
      if (is.numeric(formulas[[h]]$pitch$value)) {
        formulas[[h]]$pitch$value = exp(formulas[[h]]$pitch$value)
      }
    }
  }
  if ('formants' %in% names(f1)) {
    for (h in 1:nMorphs) {
      for (l in 1:length(f1$formants)) {
        if (is.numeric(formulas[[h]]$formants[[l]]$freq)) {
          formulas[[h]]$formants[[l]]$freq =
            exp(formulas[[h]]$formants[[l]]$freq)
        }
      }
    }
  }
  if ('formantsNoise' %in% names(f1)) {
    for (h in 1:nMorphs) {
      for (l in 1:length(f1$formantsNoise)) {
        if (is.numeric(formulas[[h]]$formantsNoise[[l]]$freq)) {
          formulas[[h]]$formantsNoise[[l]]$freq =
            exp(formulas[[h]]$formantsNoise[[l]]$freq)
        }
      }
    }
  }

  # generate the morphs based on the formulas
  for (h in 1:nMorphs) {
    sounds[[h]] = do.call(soundgen, formulas[[h]])
    if (playMorphs) playme(sounds[[h]], samplingRate = samplingRate)
    if (!is.na(savePath)){
      filename = paste0(savePath, 'morph_', h, '.wav')
      seewave::savewav(sounds[[h]], f = samplingRate, filename = filename)
    }
  }
  return(list(formulas = formulas, sounds = sounds))
}
